Searchable, Filterable List with Selection Callbacks

badranX opened the pull request:
This is related to #73. This patch adds support for filtering choices. It adds a new component, `FilterList`, instead of modifying the existing List. `FilterList` enables searching/filtering dynamically. `FilterList` also provides a callback that triggers when the user moves between choices.

Note:  I tried to minimize changes to the original library files to avoid increasing complexity. I need this for a personal project, it would be nice to just import this library.

Cube707 commented:
please rebase onto current master

Cube707 commented on examples/filter_list.py:
```diff
@@ -0,0 +1,55 @@
+import sys
+from pprint import pprint
+
+try:
+    from fuzzyfinder import fuzzyfinder
+except ImportError:
+    raise ImportError(
+        "The required package 'fuzzyfinder' is not installed. Install it using:\n\n"
+        "    pip install fuzzyfinder\n"
+    )
```
I am not the biggest fan of this. 
I get where you comming with thee examples _not_  beeing a real dependency, I would still prefere a more proper version

badranX replied:
I removed this :)

Cube707 commented on examples/filter_list.py:
```diff
@@ -0,0 +1,55 @@
+
+def filter_func(text, collection):
+    g = fuzzyfinder(text, collection, accessor=lambda x: str(x))
+    return list(g)
```
why does this require a cast to `list`?

badranX replied:
removed the cast :thumbsup:

Cube707 commented on examples/filter_list.py:
```diff
@@ -0,0 +1,55 @@
+    return list(g)
+
+
+def callback_listener(item):
+    with open('choice_logs.txt', 'a') as f:
+        f.write(str(item) + '\n')
```
I would prefer a demo to not write stuff to disk

How about updating a global var and printing that at the end?

badranX replied:
I changed it to a global var and simplified the example :thumbsup:

Cube707 commented on src/inquirer/questions.py:
```diff
@@ -76,6 +77,12 @@ def add_choice(self, choice):
             self._choices.append(choice)
             return len(self._choices) - 1
 
+    def apply_filter(self, filter_func):
+        self._filtered_choices = filter_func(self.unfiltered_choices_generator)
+
+    def remove_filter(self):
+        self._filtered_choices = None
```
these are only relevant to the new class and shouldn't be inherited by every type of question

Cube707 commented on src/inquirer/questions.py:
```diff
@@ -90,6 +97,12 @@ def default(self):
 
     @property
     def choices_generator(self):
+        choices = self._filtered_choices or self._choices
+        for choice in self._solve(choices):
+            yield (TaggedValue(*choice) if isinstance(choice, tuple) and len(choice) == 2 else choice)
+
+    @property
+    def unfiltered_choices_generator(self):
```
same here, override the property for the new class instead of adding a new, second property and adding overhead to all other classes.

badranX replied:
I isolated all filter_list features away from other classes.

Cube707 commented on src/inquirer/render/console/__init__.py:
```diff
@@ -42,6 +41,7 @@ def render(self, question, answers=None):
     def _event_loop(self, render):
         try:
             while True:
+                self.clear_eos()
```
not sure if this a unintended side effects, I will have to think about this

badranX replied:
I removed this change in the last commit. Clearing is done in _filter_list.py though now

Cube707 reviewed (changes requested):
currently broken:

- not all elements get cleared after filtering
- pressing space resets the filtering, but doesn't remove the hint

badranX commented:
Thanks for the insightful review! I'll get back to this PR as soon as possible.

badranX commented:
I isolated filter_list features away from everything else like you asked. I hope it's ok like this. I'll write tests/docs in my next free time :sweat_smile:

badranX commented:
- Removed the "Selection Callbacks" part. I think it's not important. 
 - Overloaded the single FilterList example with tag/hints/carousel/other features.
 - Added 0 change to the original behaviour of the library, this PR can be reverted easily if proved hard to maintain.
 
Please tell me if docs are needed or any other changes. Since it behaves similarly to List, I think the example is enough.
